### PR TITLE
Enable workload identity authentication for the Databricks provider

### LINKED ISSUE #41586: Support for workload identity authentication from the Databricks provider
### Description

Currently, only managed identity is supported for connecting from Airflow to Databricks using the `use_azure_managed_identity` parameter. However, I have a use case where an Airflow instance is running on an AKS cluster. That's why this would be a great improvement in authentication without using any secrets.

### Use case/motivation

_No response_

### Related issues

#39521 

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by basvandriel**
This pull request adds support for [authenticating using Workload Identity ](https://learn.microsoft.com/nl-nl/azure/aks/workload-identity-overview?tabs=dotnet)for the Databricks provider. In the provided unit tests, this has been tested using mocking the token generation. However, the `DefaultAzureCredential().get_token(...)` part has been tested on an actual AKS cluster. This closes #41586.

---

### CODE REVIEW

File: airflow/providers/databricks/hooks/databricks_base.py
Comment (jrderuiter) [inline airflow/providers/databricks/hooks/databricks_base.py]: In principle getting a token in this manner isn't restricted to workload identity perse. For example, using the `DefaultAzureCredential` you can support any of the authentication options listed here: https://learn.microsoft.com/en-us/dotnet/azure/sdk/authentication/credential-chains?tabs=dac#defaultazurecredential-overview. 

Would it make sense to rename this method to reflect this and maybe use an options such as `use_azure_default_credential` instead? 

For local development this more general option would also be awesome, as it (for example) allows you to use your personal Entra ID together with the Azure CLI to authenticate to a Databricks cluster.

Comment (basvandriel) [reply airflow/providers/databricks/hooks/databricks_base.py]: Great suggestion. The change has been made!

**Comment by BasPH** [comment]
Looks okay to me, although there's quite a lot of code duplication between the `_get_aad_token_for_default_az_credential` and `_a _get_aad_token_for_default_az_credential` methods. Any way to extract common code?

**Comment by jrderuiter** [comment]
This duplication is essentially following the existing pattern that's already in place in the hook (if you look at the other methods, there is very often both a sync and an async implementation). 

To reduce duplication we could break out the setup (lines 430-434) and the checking of the token (lines 450-456) into separate functions but I'm not sure if that would really improve readability. 

A more heavy-weight alternative would be to only define the async version and then define the sync version using [asyncio.run](https://docs.python.org/3/library/asyncio-runner.html#asyncio.run):

```
def _get_aad_token_for_default_az_credential(self, resource: str) -> str:
    return asyncio.run(self. _a_get_aad_token_for_default_az_credential(resource)
```

However, this would add some overhead to the sync function call as `asyncio.run` starts an async event loop in the background:

> This function runs the passed coroutine, taking care of managing the asyncio event loop, finalizing asynchronous generators, and closing the executor.

In short, to match the existing style I would prefer leaving it as is. However, I'm happy to explore these other options if needed. Also open to any suggestions from others :)

**Comment by jrderuiter** [comment]
@basvandriel Can you give me write access to your PR fork + branch? Then I can take over this PR as discussed privately.

**Comment by basvandriel** [comment]
Hi @jrderuiter, I was already working on this. I will finalize this today. Keep you posted!

**Comment by BasPH** [comment]
@jrderuiter In that case keeping the duplicated code is okay with me. Could you fix the static code checks?

**Comment by basvandriel** [comment]
While I'm running my pre-commit checks I made some update on the duplicated code. Unfortunately, using `asyncio.run` is not an option. The code that is currently is being used uses a different SDK for the asynchronous part of the token generation.

However, I think the readability increases by having a `Callable` argument, where a function returns an `AccessToken`. This way we only have 2 functions instead of 4, one for the asynchronous one and one the synchronous.

File: airflow/providers/databricks/hooks/databricks_base.py
Comment (jrderuiter) [inline airflow/providers/databricks/hooks/databricks_base.py]: I'd change this message to indicate that you're using the AzureDefaultCredential instead of referencing workload identity. Workload identity is only one of the (many) mechanisms that the AzureDefaultCredential can use for authentication and we have no guarantee that it will actually use workload identity.

File: airflow/providers/databricks/hooks/databricks_base.py
Comment (jrderuiter) [inline airflow/providers/databricks/hooks/databricks_base.py]: I'd remove this check, as using workload identity from Kubernetes is only one of the options that the AzureDefaultCredential can use for authentication. We should just use the AzureDefaultCredential and let it figure out which mode it will use for authentication.

File: airflow/providers/databricks/hooks/databricks_base.py
Comment (jrderuiter) [inline airflow/providers/databricks/hooks/databricks_base.py]: I'd use a different term than `executor` here, as `executor` has a specific meaning in the Airflow context: https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/executor/index.html.

File: airflow/providers/databricks/hooks/databricks_base.py
Comment (jrderuiter) [inline airflow/providers/databricks/hooks/databricks_base.py]: I still see a lot of code duplication here. Aren't we just making it more complicated by adding this executor function if it only saves us 1-2 lines of duplication?

**Review by jrderuiter — COMMENTED** [review]
I'm not really sure about the refactoring TBH, now it feels like we add an extra layer of complexity (the `executor` functions + `get_aad_token` functions) with minimal reduction in duplication.

File: airflow/providers/databricks/hooks/databricks_base.py
Comment (basvandriel) [reply airflow/providers/databricks/hooks/databricks_base.py]: You are right. If we want a similar approach, we would need 2 functions for the wrapping code and the rest for the variations of retrieving the token. I'm going to revert my commit and apply the other suggestions. Thanks!

**Comment by basvandriel** [comment]
> I'm not really sure about the refactoring TBH, now it feels like we add an extra layer of complexity (the `executor` functions + `get_aad_token` functions) with minimal reduction in duplication.

You are right. If we want a similar approach, we would need 2 functions for the wrapping code and the rest for the variations of retrieving the token. I'm going to revert my commit and apply the other suggestions. Thanks!

File: airflow/providers/databricks/hooks/databricks_base.py
Comment (jrderuiter) [inline airflow/providers/databricks/hooks/databricks_base.py]: Small typo in the log:  `AzureDefaultCredentialy` -> `AzureDefaultCredential`.

File: airflow/providers/databricks/hooks/databricks_base.py
Comment (jrderuiter) [inline airflow/providers/databricks/hooks/databricks_base.py]: This key should  also be added to the `extra_parameters` list in line 89 I think.

**Review by jrderuiter — COMMENTED** [review]
LGTM @basvandriel , two small comments.

**Comment by jrderuiter** [comment]
I see some of the static checks are failing, can you have a look at this @basvandriel? Afterwards we should be fine to merge if @BasPH agrees.

File: airflow/providers/databricks/hooks/databricks_base.py
Comment (BasPH) [inline airflow/providers/databricks/hooks/databricks_base.py]: Would link to the English docs:

```suggestion
                    # While there is a WorkloadIdentityCredential class, the below class is advised by Microsoft
                    # https://learn.microsoft.com/en-us/azure/aks/workload-identity-overview
```

File: airflow/providers/databricks/hooks/databricks_base.py
Comment (BasPH) [inline airflow/providers/databricks/hooks/databricks_base.py]: Remove unnecessary comment

```suggestion
```

File: airflow/providers/databricks/hooks/databricks_base.py
Comment (BasPH) [inline airflow/providers/databricks/hooks/databricks_base.py]:
```suggestion
                    # While there is a WorkloadIdentityCredential class, the below class is advised by Microsoft
                    # https://learn.microsoft.com/en-us/azure/aks/workload-identity-overview
```

**Review by BasPH — COMMENTED** [review]
Did another review, few nitty comments but otherwise LGTM

**Comment by jrderuiter** [comment]
@basvandriel Do you need any help fixing the static checks?

**Comment by basvandriel** [comment]
> @basvandriel Do you need any help fixing the static checks?

Hi @jrderuiter, I'm sorry for the late reply. I will be looking into this today.

**Comment by BasPH** [comment]
Hi @basvandriel almost looks okay to me. There's one tiny static check nit that thinks that "aks" is a typo: https://github.com/apache/airflow/actions/runs/10882398151/job/30464268761?pr=41639#step:8:80. Could you add "aks" to `docs/spelling_wordlist.txt`?

The other CICD failure (https://github.com/apache/airflow/actions/runs/10882398151/job/30464269767?pr=41639) is one I cannot explain at first sight. Will need to deepdive to understand what's failing there.

**Review by BasPH — APPROVED** [review]

**Comment by BasPH** [comment]
All green now, it took a while but thanks for hanging on @basvandriel @jrderuiter!

**Pull request merged**
